### ISSUE TITLE
non-manual foldmethod makes botline incorrect in decoration provider during :edit-ing existing buffer

### ISSUE THREAD

**Issue by tomtomjhj** [neovim/neovim#28443]
### Problem

* If `foldmethod` is set to anything other than `manual`,
* when `:edit`-ing an existing buffer,
* in decoration provider `on_win` callback,
* the `botline` parameter (or `vim.fn.line('w$')`) is wrong.

This may result in wrong treesitter highlighting for injections. But this is not very severe problem since redraw will be triggered with correct botline at some point. 

### Steps to reproduce

Add the following test case

```diff
diff --git a/test/functional/ui/decorations_spec.lua b/test/functional/ui/decorations_spec.lua
index eea62140c..c864df8df 100644
--- a/test/functional/ui/decorations_spec.lua
+++ b/test/functional/ui/decorations_spec.lua
@@ -785,6 +785,36 @@ describe('decorations providers', function()
     ]])
     eq(2, exec_lua([[return _G.cnt]]))
   end)
+
+  it('provides correct botline regardless of foldmethod', function()
+    command('set foldmethod=expr') -- any non-manual fdm
+
+    local file = t.tmpname()
+    t.write_file(file, mulholland)
+    command('edit ' .. file)
+    command('enew')
+
+    exec_lua('_G.log = {}')
+    setup_provider([[
+      function on_do(event, win, buf, topline, botline)
+        if event == 'win' and buf == 1 then
+          table.insert(_G.log, {topline, botline})
+        end
+      end
+    ]])
+
+    command('buffer 1')
+    feed('<C-^>')
+    command('edit ' .. file)
+    command('redraw!')
+    n.poke_eventloop()
+
+    eq({
+      { 0, 6 }, -- :buffer
+      { 0, 6 }, -- :edit
+      { 0, 6 }, -- :redraw!
+    }, exec_lua('return _G.log'))
+  end)
 end)
 
 local example_text = [[
```

Run the test (`make TEST_FILE=test/functional/ui/decorations_spec.lua make functionaltest`).

Result
```
Passed in:
(table: 0x7c12c74fc3e8) {
  [1] = {
    [1] = 0
    [2] = 6 }
 *[2] = {
    [1] = 0
   *[2] = 1 }
  [3] = {
    [1] = 0
    [2] = 6 } }
Expected:
(table: 0x7c12c75c3238) {
  [1] = {
    [1] = 0
    [2] = 6 }
 *[2] = {
    [1] = 0
   *[2] = 6 }
  [3] = {
    [1] = 0
    [2] = 6 } }
```

The correct value is passed for `:buffer`, but not for `:edit`. 

### Expected behavior

The test should pass.

### Neovim version (nvim -v)

0.10 032df963bb3fb0b5652e1817e9f4da986996fa6d

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

ubuntu 22.04

### Terminal name/version

GNOME Terminal 3.44.0 using VTE 0.68.0 +BIDI +GNUTLS +ICU +SYSTEMD

### $TERM environment variable

tmux-256color

### Installation

build from repo

**Comment by tomtomjhj** [comment]
This fixes this issue. 
```diff
diff --git a/src/nvim/window.c b/src/nvim/window.c
index 8232767e2..1a6c3f726 100644
--- a/src/nvim/window.c
+++ b/src/nvim/window.c
@@ -2467,6 +2467,7 @@ void win_init_empty(win_T *wp)
   wp->w_topline = 1;
   wp->w_topfill = 0;
   wp->w_botline = 2;
+  wp->w_valid = 0;
   wp->w_s = &wp->w_buffer->b_s;
 }
 
```
`win_init_empty` is invoked for the current existing window when running [`:edit`](https://github.com/neovim/neovim/blob/47dbda97d2f40729733b1c0d1d13d914065af23c/src/nvim/ex_cmds.c#L2552), whose `w_valid` may have `VALID_BOTLINE`. Resetting `w_botline` without marking it as invalid makes `validate_botline` in `decor_providers_invoke_win` no-op, resulting in wrong `botline` in `on_win`.

Vim has the same issue, but I'm not sure if this can manifest as a user-observable bug. I'll think about it and send a patch to Vim.

**Comment by tomtomjhj** [comment]
Fixed by #28530
